pkg/scheduler/framework/runtime/framework.go

```diff
@@ -526,7 +527,7 @@ func (f *frameworkImpl) expandMultiPointPlugins(logger klog.Logger, profile *con
 		// - part 3: other plugins (excluded by part 1 & 2) in regular extension point.
 		newPlugins := reflect.New(reflect.TypeOf(e.slicePtr).Elem()).Elem()
 		// part 1
-		for _, name := range enabledSet.list {
+		for _, name := range slice.CopyStrings(enabledSet.list) {
```

Comment:
the root cause is that during orderedset.delete(), the length of the list get changed, so some item missed its chance to be iterated in the outer for loop.

The fix can be snapshotting this list in the outer for loop - as you're proposing here. Could you help check if slice.CopyStrings is available in all supported versions 1.25? ~ 1.28? as we want to cherrypick this fix.

Or, another fix is to make the for loop logic accurate w/o copying/snapshotting the list:

```diff
diff --git a/pkg/scheduler/framework/runtime/framework.go b/pkg/scheduler/framework/runtime/framework.go
index 8caa8ec1cb4..f238bcbdfb7 100644
--- a/pkg/scheduler/framework/runtime/framework.go
+++ b/pkg/scheduler/framework/runtime/framework.go
@@ -450,12 +450,10 @@ func (os *orderedSet) has(s string) bool {
 	return found
 }
 
-func (os *orderedSet) delete(s string) {
-	if i, found := os.set[s]; found {
-		delete(os.set, s)
-		os.list = append(os.list[:i-os.deletionCnt], os.list[i+1-os.deletionCnt:]...)
-		os.deletionCnt++
-	}
+func (os *orderedSet) delete(i int) {
+	delete(os.set, os.list[i])
+	os.list = append(os.list[:i], os.list[i+1:]...)
+	os.deletionCnt++
 }
 
 func (f *frameworkImpl) expandMultiPointPlugins(logger klog.Logger, profile *config.KubeSchedulerProfile, pluginsMap map[string]framework.Plugin) error {
@@ -526,10 +524,12 @@ func (f *frameworkImpl) expandMultiPointPlugins(logger klog.Logger, profile *con
 		// - part 3: other plugins (excluded by part 1 & 2) in regular extension point.
 		newPlugins := reflect.New(reflect.TypeOf(e.slicePtr).Elem()).Elem()
 		// part 1
-		for _, name := range enabledSet.list {
+		for i := range enabledSet.list {
+			j := i - enabledSet.deletionCnt
+			name := enabledSet.list[j]
 			if overridePlugins.has(name) {
 				newPlugins = reflect.Append(newPlugins, reflect.ValueOf(pluginsMap[name]))
-				enabledSet.delete(name)
+				enabledSet.delete(j)
 			}
 		}
 		// part 2
```

Reply:
Thank you for the detailed explanation! Yes, git blame shows that the code for `slice.CopyStrings` has not been modified in almost 6 years. I have tested this fix on our 1.24 cluster and it works fine.
![image](https://github.com/kubernetes/kubernetes/assets/18328639/abbcd0a5-7f66-46e7-8593-0ecb5e968cb5)

Reply:
thanks for confirming. Could you help cherrypick it to all supported lower k8s versions?

Reply:
Sure, I will cherrypick it after this PR is merged.

Reply:
Hi, I have cherry-picked it to 1.26-1.29. Could you PTAL?
#122366
#122368
#122370
#122371